### ISSUE TITLE
Concurrency problem with Serialising, Polymorphism (2.7.0+)

### ISSUE THREAD

**Issue by afmm** [FasterXML/jackson-databind#1579]
Greetings,

We're having an unusual bug occur that we're unable to create a reproduction for outside of our own application code, so hopefully a description will do for now. 

We've been trying to upgrade Jackson from 2.6.7 up to 2.7+, including the latest 2.8.7. A consistent problem we've had since 2.7.0 is that when serialising an object with multiple levels of inheritance, it doesn't proceed up the full inheritance tree in one case. 

For example:

Interface A
Abstract Class B<T extends B> implements A
Interface C
Abstract Class D extends B implements C
Concrete Class E extends D
Concrete Class F extends D

When serialising E, it only includes members from E & D. 
When serialising F, it includes members from F, D & B. 

We're reliably able to reproduce this issue on our build servers (Jenkins, running on Linux Slaves) but only for builds that are started via Jenkins. If we run the build manually (mvn clean install) on a Slave, it passes. 
We've noticed the same issue intermittently in our production like environments, however we've never been able to reproduce the issue on a local dev machine. 

Does anyone have any suggestions as to how we may be able to reproduce this? 

We're also quite happy to try new versions, even directly from GitHub.

Thanks!

**Comment by cowtowncoder** [comment]
That problem does not ring a bell unfortunately. If I had to speculate, I would suspect it could have something to do with caching done by type resolution, although I haven't observed anything like this or seen reproducible issue reports.
It is too bad that this occurs with even 2.8.7 since there is one fix there that could have been relevant. 2.8.8 (to be released) does not have -- as far as I know -- anything that should change introspection.

One thing you could try if is easy enough to do would be to try to out latest pre-release version, 2.9.0.pr2. While you wouldn't use it for production, I would be interested in seeing whether it would have any differences compared to 2.8.

As to 2.6.7 -> 2.7.0: this makes some sense in that 2.7 contained full rewrite of type resolution system, and while it solved all remaining issues regarding resolution of type variables, it also adds some more caching. I didn't think that could cause issues, but now I am wondering if there could be some problems regarding access to JVM/JDK accessor for generic type information.

A reliably way to reproduce issues would obviously be great, since I could play with various things, essentially disabling some aspects of caching, and/or adding synchronization.

I will try to go over code in areas that I think are relevant: this is mostly accessed via `TypeFactory`, code in package `com.fasterxml.jackson.databind.introspect`: especially `AnnotatedClass`. But I have to admit I do not have good candidates for root cause.

**Comment by cowtowncoder** [comment]
Some more questions:

1. Which JVM versions is the problem observed with? (and if there are others where it doesn't occur, version of those)
2. Of sort of advanced features, like say -- object id, polymorphic handling, `@JsonUnwrapped`, builders -- are there some that would seem to be correlated (used for POJOs that are affected)?

I am guessing that aside from potential for type resolution to be affected, there could be other explanations too, side effects of unintended modifications for unwrapping, or for property ignorals with `@JsonIgnoreProperties` applied on a property.

**Comment by afmm** [comment]
Hi @cowtowncoder, thanks for your quick response.

I've tried 2.9.0.pr2 and we're seeing the same issue, also seeing a lot of other tests break which are probably migration related things we will need to look at when the time comes.

With regards to the JVM versions where we've observed it - 
```
openjdk version "1.8.0_121"
OpenJDK Runtime Environment (build 1.8.0_121-b13)
OpenJDK 64-Bit Server VM (build 25.121-b13, mixed mode)
```
We've also seen it work fine on this version too.

With regards to Advanced features, this is the configuration for our ObjectMapper:

```
    ObjectMapperFactoryImpl() {
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.registerModule(new JodaModule());
        objectMapper.registerModule(new GuavaModule());
        objectMapper.registerModule(new Jdk8Module());
        objectMapper.registerModule(new AfterburnerModule());
        loadJavaDateTimeModule(objectMapper);
        objectMapper.configure(com.fasterxml.jackson.databind.SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
        objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);

        ObjectMapper objectMapperTypeClient = objectMapper.copy();
        objectMapperTypeClient.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

        objectMappers = ImmutableMap.of(JsonConfigType.Types.DEFAULT, objectMapper, JsonConfigType.Types.CLIENT, objectMapperTypeClient);
    }

    /**
     * In jackson-datatype-jsr310-2.4.x the module is called {@link com.fasterxml.jackson.datatype.jsr310.JSR310Module} whereas in
     * jackson-datatype-jsr310-2.6.X that class got deprecated and we should use {@link com.fasterxml.jackson.datatype.jsr310.JavaTimeModule}
     *
     * @param mapper the jackson mapper being configured
     */
    private void loadJavaDateTimeModule(ObjectMapper mapper) {
        try {
            Class<?> newerDateTimeModule = Class.forName("com.fasterxml.jackson.datatype.jsr310.JavaTimeModule");
            mapper.registerModule((Module) newerDateTimeModule.newInstance());
        } catch (ClassNotFoundException e) {
            //class not found, we try to load the old module
            try {
                Class<?> olderDateTimeModule = Class.forName("com.fasterxml.jackson.datatype.jsr310.JSR310Module");
                mapper.registerModule((Module) olderDateTimeModule.newInstance());
            } catch (Exception e2) {
                throw new RuntimeException("Could not load JSR310Module module. Is jackson-datatype-jsr310 in your classpath?");
            }
        } catch (Exception e) {
            throw new RuntimeException("Could not instantiate JavaTimeModule module.");
        }
    }
```
The type information is defined in the interface, A:

```
@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "messageType"
)
@JsonSubTypes({
        @JsonSubTypes.Type(value = E.class, name = A.E_ID),
        @JsonSubTypes.Type(value = F.class, name = A.F_ID),
})
```

The property ```messageType``` is provided in the constructor for the concrete class, e.g., E or F, but is a member of B, and is populated via ```super()``` calls from the subclasses.

If you have an idea as to what may make it reproducible or any configuration options we can tweak, I'm happy to test, however we're only seeing it occur reliably on a single test under the above mentioned circumstances. We've tried to create an isolated example of the code in question but we haven't been able to recreate the same issue with it.

**Comment by cowtowncoder** [comment]
@afmm Ok, nothing too fancy there. I would probably try disabling Afterburner since it plugs in at quite low level so perhaps it could somehow interfere (although I have no idea how that would be).
And although number of datatype modules is non-trivial none of them really should do anything particularly active so I am not too worried there.

As to JVM version I still build with 1.7, but wrt 1.8 I use very close version (1.8.0.122-ea). But then again none of the tests are heavily multi-threader or use combination of multiple modules.
So I would not necessarily observe this during development.

I have not seen anything like this in production but version we use mostly is 2.7.9 (due to some Spark limitations).
This doesn't really tell much, just that I am not familiar with possible work-arounds not having to try any. There have occasionally been reports about suspected issues with concurrency but even there these symptoms sound quite different; the only other open investigation has to do with odd casing (capitalized vs lower-cased) regarding symbol matching.

**Comment by cowtowncoder** [comment]
Going back to the case itself: would it be possible to include minimal class hierarchy with which problem has been observed? By minimal I only mean "as small as practical to prune to", dropping fields/getters. I realize this may not be easy to do but... well would be great to have. Even if I could not reproduce symptom, as long as I could see how classes are defined it might provide more clues (above and beyond initial explanation which was complete enough -- but sometimes it helps to see properties too, not just inheritance).

**Comment by cowtowncoder** [comment]
Hmmh. I have a suspicion that this _could_ be due to possible problem with combination of lazy evaluation and caching of `AnnotatedClass`: there could be a that partially resolved set of methods (getters, setters) could be briefly visible. This did not matter earlier when `AnnotatedClass` instances weren't cached... although, caching seems to have been added in 2.5 so problem should go back that far.

I am not sure this is what is causing the issue but seems like about only realistic lead. Too bad I just released 2.8.8.

**Comment by afmm** [comment]
Hi @cowtowncoder , 

Is there anything we can try to test the above suspicion? I'm working on cutting down the classes to a minimal set that reproduces the problem at the moment, but if there are some settings we can try to the object mapper instead that would probably be a lot easier.

**Comment by cowtowncoder** [comment]
@afmm One that could help is if it is possible to try `2.9.0.pr3`: handling of not only `AnnotatedClass` but its content (fields, methods, constructors) collection has been rewritten and should be much safer.

**Comment by afmm** [comment]
Hi @cowtowncoder , no luck with that version of Jackson, I've had some time to sit down and prepare some code. The attached are our model objects which have this issue, and they are replicating the bug at this point with these classes, in Jenkins builds.

```
Unrecognized field "aseven" (class model.E), not marked as ignorable (10 known properties: "dOne", "aFive", "cfive", "aFour", "eThree", "cthree", "gOne", "cone", "ctwo", "cfour"])
 at [Source: (byte[])"{"eThree":false,"aseven":"4.22","cfive":"Foo","cone":"Bar","ctwo":"foo-bar","cthree":"bar-foo"}"; line: 1, column: 108] (through reference chain: model.E["aseven"])
```

[jackson-bug.tar.gz](https://github.com/FasterXML/jackson-databind/files/1070398/jackson-bug.tar.gz)

**Comment by afmm** [comment]
I've tested the newest release, 2.9.0 and that seems to have fixed our issue. 

I've confirmed that the issue is still present in our current code base with 2.8.6, and resolved in 2.9.0. 

I'm happy to close this issue out as fixed, thanks for the help!

**Comment by cowtowncoder** [comment]
@afmm Thank you for verifying this. This is good news.

I spent quite a bit of time to refactor how annotations are handled (`AnnotedClass` and its interactions), to try to make sure that concurrency is handled safely, so I am happy that this work finally improves things. Too bad it took this long to find both problems and solutions; I assume this has been the case for all preceding 2.x versions (and actually probably 1.x as well).
Some (but not all) of changes are also in `2.8.9`, I think, but if at all possible I think going to `2.9.0` makes sense.
